server/src/main/java/org/apache/lucene/index/BaseStarTreeBuilder.java

```diff
@@ -0,0 +1,647 @@
+                    numericType = ((NumberFieldMapper) fieldMapper).fieldType().numericType();
+                } else {
+                    logger.error("unsupported mapper type");
+                    throw new IllegalStateException("unsupported mapper type");
+                }
+
+                FieldInfo metricFieldInfos = state.fieldInfos.fieldInfo(metric.getField());
+                DocValuesType metricDocValuesType = metricFieldInfos.getDocValuesType();
+                if (metricStat != MetricStat.COUNT) {
+                    metricStatReader = starTreeDocValuesIteratorAdapter.getDocValuesIterator(
+                        metricDocValuesType,
+                        metricFieldInfos,
+                        fieldProducerMap.get(metricFieldInfos.name)
+                    );
+                } else {
+                    metricStatReader = new SequentialDocValuesIterator();
+                }
+
+                MetricAggregatorInfo metricAggregatorInfo = new MetricAggregatorInfo(
+                    metricStat,
+                    metric.getField(),
+                    starTreeField.getName(),
+                    numericType,
+                    metricStatReader
+                );
+                metricAggregatorInfos.add(metricAggregatorInfo);
+            }
+        }
+        return metricAggregatorInfos;
+    }
+
+    /**
+     * Adds a document to the star-tree.
+     *
+     * @param starTreeDocument star tree document to be added
+     * @throws IOException if an I/O error occurs while adding the document
+     */
+    public abstract void appendStarTreeDocument(StarTreeDocument starTreeDocument) throws IOException;
+
+    /**
+     * Returns the document of the given document id in the star-tree.
+     *
+     * @param docId document id
+     * @return star tree document
+     * @throws IOException if an I/O error occurs while fetching the star-tree document
+     */
+    public abstract StarTreeDocument getStarTreeDocument(int docId) throws IOException;
+
+    /**
+     * Retrieves the list of star-tree documents in the star-tree.
+     *
+     * @return Star tree documents
+     */
+    public abstract List<StarTreeDocument> getStarTreeDocuments();
+
+    /**
+     * Returns the value of the dimension for the given dimension id and document in the star-tree.
+     *
+     * @param docId       document id
+     * @param dimensionId dimension id
+     * @return dimension value
+     */
+    public abstract Long getDimensionValue(int docId, int dimensionId) throws IOException;
+
+    /**
+     * Sorts and aggregates the star-tree document in the segment, and returns a star-tree document iterator for all the
+     * aggregated star-tree document.
+     *
+     * @param numDocs number of documents in the given segment
+     * @return Iterator for the aggregated star-tree document
+     */
+    public abstract Iterator<StarTreeDocument> sortAndAggregateStarTreeDocuments(int numDocs) throws IOException;
```

Comment:
do you mean numDocs is not required? We can actually make do without it because ```totalSegmentDocs``` is protected, and hence available in the OnHeap / OffHeap implementation. 
Lmk, can remove this.